File: samples/leetcode/sliding_window/best_time_to_buy_a_stock.rs
use std::error::Error;
use std::cmp::max;

struct Solution;

impl Solution {
    pub fn max_profit(prices: Vec<i32>) -> i32 {
        let mut buy_index: usize = 0;
        let mut max_profit = 0;

        for sell_index in 1..prices.len() {
            if prices[sell_index] < prices[buy_index] {
                buy_index = sell_index;
            }
            max_profit = max(max_profit, prices[sell_index] - prices[buy_index]);
        }
        max_profit
    }
}

fn main() -> Result<(), Box<dyn Error>> {

    let vec = vec![7,1,5,3,6,4];
    println!("{}", Solution::max_profit(vec));

    Ok(())
}

File: bazel/samples/leetcode/sliding_window/best_time_to_buy_a_stock.cpp
#include <vector>
#include <iostream>

using namespace std;

class Solution {
public:
    int maxProfit(vector<int>& prices) {
        size_t buyIndex = 0;
        int maxProfit = 0;
        for (size_t sellIndex = 1; sellIndex < prices.size(); sellIndex++) {
            if (prices[sellIndex] < prices[buyIndex]) {
                buyIndex = sellIndex;
                continue;
            }
            maxProfit = std::max(maxProfit, prices[sellIndex] - prices[buyIndex]);
        }
        return maxProfit;
    }
};

int main() {
    /**
    Input: prices = [7,1,5,3,6,4]
    Output: 5
    Explanation: Buy on day 2 (price = 1) and sell on day 5 (price = 6), profit = 6-1 = 5.
    Note that buying on day 2 and selling on day 1 is not allowed because you must buy before you sell.
     */

    auto solution = Solution();
    auto prices = vector<int>{7,1,5,3,6,4};
    std::cout << solution.maxProfit(prices) << std::endl;
    return 0;
}
